cpp/src/text/subword/subword_tokenize.cu

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/utilities/error.hpp>
+#include <nvtext/detail/load_hash_file.hpp>
+#include <nvtext/subword_tokenize.hpp>
+#include <text/subword/detail/wordpiece_tokenizer.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/transform_scan.h>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+/**
+ * @brief Convert tokens and row2log map to final tensor data.
+ *
+ * @param[in] token_ids Tokens from tokenizer
+ * @param[in] offsets Offsets to each string's output row of tokens
+ * @param[in] row2log String to log token counts
+ * @param[in] row2row_within_log Token counts within sub-rows of the output
+ * @param[in] max_sequence_length Maximum number of tokens in a row
+ * @param[in] stride Number of tokens in sub-rows
+ * @param[in] do_truncate True if tokens should not spill into sub-rows in the output
+ * @param[out] final_tensor Output vector of token-ids
+ * @param[out] attn_mask Identifies valid token id entries
+ * @param[out] metadata Additional data per row
+ */
+__global__ void kernel_compute_tensor_metadata(
+  // input
+  uint32_t const* token_ids,
+  uint32_t const* offsets,
+  uint32_t const* row2log,
```

Comment:
What is the log here? Is this a vestige of this codes usage for log parsing?

Reply:
This is the final output format. This appears to be called a tensor so I can change `log` to `tensor`.